### PR TITLE
Filtering for eth_getLogs

### PR DESCRIPTION
**Pull request by dutterbutter**
Changes made in this PR:
- Added `PublicFilterAPI`
- Added Logs query in `querier.go`
- Added `filterLogs`, `UnmarshalJSON` funcs from `go-ethereum` 
- Added helper funcs

To Test:

```
make install 
rm -rf ~/.emint*
emintd init moniker --chain-id 3
emintcli config chain-id 3
emintcli config output json
emintcli config indent true
emintcli config trust-node true
emintcli keys add dustin
testpass
testpass
emintcli emintkeys add dustineth
testpass
testpass
emintd add-genesis-account $(emintcli keys show dustin -a) 1000photon,100000000stake
emintd add-genesis-account $(emintcli emintkeys show dustineth -a) 100000000photon,100000000stake
emintd gentx --name dustin
testpass
emintd collect-gentxs
emintd validate-genesis
emintd start --pruning=nothing --rpc.unsafe --log_level "main:info,state:info,mempool:i
```

> Start server with unlocked key to send tx:
```
emintcli rest-server --laddr "tcp://localhost:8545" --unlock-key dustineth
```

Deploy Contract and Emit Events:
> While ethermint node is running:
```
git clone git@github.com:ChainSafe/ethermint-scripts.git
cd web3
node test_contract.js
```

> Get logs with a range of blocks
```
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"fromBlock":"0x7", "toBlock: "0x9"}],"id":6}' -H "Content-Type: application/json" http://localhost:8545/rpc
```
> Get logs with address criteria (this can also take an array of addresses)
```
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"address":"<contract_address"}],"id":6}' -H "Content-Type: application/json" http://localhost:8545/rpc
```

Remaining todos for logs:
- Need to set `blockHash` in `Prepare` func prior to log entry (will also then enable filtering by `blockHash`
- Subscribing to event logs
- Using the bloom filter to optimize log querying 
- additional filter rpc calls

### CODE REVIEW

File: rpc/filter_api.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	"math/big"
+
+	eth "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type FilterCriteria eth.FilterQuery
```
Comment (austinabell) [inline rpc/filter_api.go]: What is the reason that Ethereum's type could not be used as is? FilterCriteria type is not in internal package and seems unmodified

File: rpc/filter_api.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		if criteria.FromBlock != nil {
+			begin = criteria.FromBlock.Int64()
+		}
+		from := big.NewInt(begin)
+		end := rpc.LatestBlockNumber.Int64()
+		if criteria.ToBlock != nil {
+			end = criteria.ToBlock.Int64()
+		}
+		to := big.NewInt(end)
+		results := e.getLogs()
```
Comment (austinabell) [inline rpc/filter_api.go:54]: Would it not be beneficial to include the range filter on the type you created to abstract this logic so you aren't doing it duplicate times?

File: rpc/filter_api.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	}
+	return logs
+}
+
+/*
+UnmarshalJSON taken from go-ethereum
+*/
+
+// UnmarshalJSON sets *args fields with given data.
+func (args *FilterCriteria) UnmarshalJSON(data []byte) error {
```
Comment (austinabell) [inline rpc/filter_api.go]: Code is identical to geth's FilterCriteria, don't think it's needed to duplicate?

File: rpc/filters.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+/*
+	- Filter functions derived from go-ethereum
+	Used to set the criteria passed in from RPC params
+*/
+
+// Filter can be used to retrieve and filter logs.
+type Filter struct {
```
Comment (austinabell) [inline rpc/filters.go:16]: Reason for excluding range filter parameters and bloom filter matcher? just leaving out just for this PR to be added later or is there another reason?

**Review by austinabell — CHANGES_REQUESTED** [review]

File: rpc/filter_api.go (hunk 1 above)
Comment (dutterbutter) [reply rpc/filter_api.go]: You can only define methods on a type defined in that same package. So the `UnmarshalJSON` receiver function that requires `eth.FilterQuery` can be used appropriately within the same package.

File: rpc/filter_api.go (hunk 3 above)
Comment (dutterbutter) [reply rpc/filter_api.go]: see above comment

File: rpc/filters.go (hunk 4 above)
Comment (dutterbutter) [reply rpc/filters.go:16]: Exactly, it will be added to in subsequent PR that makes use of the bloom filter matcher...the linter was just complaining during the `structcheck` that these fields were unused.

File: rpc/filter_api.go (hunk 2 above)
Comment (dutterbutter) [reply rpc/filter_api.go:54]: can you elaborate on what you mean from the above?

File: rpc/filter_api.go (hunk 1 above)
Comment (austinabell) [reply rpc/filter_api.go]: I'm saying to just use Ethereum's type. It's only used for unmarshalling to be used and you don't introduce any new functionality in this PR. My suggestion I just did in a commit here: https://github.com/ChainSafe/ethermint/commit/a2dfd0bfafed74f70d0162ab6fe10061809aa281 let me know why this doesn't work for what you're doing

File: rpc/filter_api.go (hunk 2 above)
Comment (austinabell) [reply rpc/filter_api.go:54]: Just replicating the e.getLogs and filterLogs calls, can't the filter just be generated in this if/ else then call these equivalently outside of it for consistency?

File: rpc/filters.go (hunk 4 above)
Comment (austinabell) [reply rpc/filters.go:16]: Gotcha! Just was curious since the range filter functionality was somewhat included in this PR (Although maybe not the focus) so as long as we remember to come back to this

File: rpc/filter_api.go (hunk 1 above)
Comment (dutterbutter) [reply rpc/filter_api.go]: I see! Yeah, miscommunication there. For sure will make the change.

File: rpc/filter_api.go (hunk 2 above)
Comment (dutterbutter) [reply rpc/filter_api.go:54]: still need some clarity here.... `e.getLogs` is the actual call to retrieve the logs whereas `filterLogs` does the filtering. We want to keep this functionality separate when we introduce bloom filters so that we only call `filterLogs` on the logs that _may_  meet the filter criteria which will be indicated by the bloom filter.

Comment (austinabell) [reply rpc/filter_api.go:54]: to be more specific, I just mean having the functionality to create a range filter like you do with `NewBlockFilter()` then get logs and filter them below this conditional (let me know what you think of this

Comment (austinabell) [reply rpc/filter_api.go:54]: The only problem I have is that you are handling block filters and range filters asymmetrically, which is somewhat hard to follow. Also small thing with this is that it adds a linting issue with returning both in the coniditonal (that doesn't matter as much)

**Review by austinabell — APPROVED** [review]
:D
